nexus/db-queries/src/db/datastore/db_metadata.rs

```diff
@@ -8,21 +8,183 @@ use super::DataStore;
+        {
+            if entry.file_type().await.map_err(|e| e.to_string())?.is_dir() {
+                let name = entry
+                    .file_name()
+                    .into_string()
+                    .map_err(|_| "Non-unicode schema dir".to_string())?;
+                if let Ok(observed_version) = name.parse::<SemverVersion>() {
+                    all_versions.insert(observed_version);
+                } else {
+                    warn!(log, "Failed to parse {name} as a semver version");
+                }
+            }
+        }
+
+        if all_versions.get(&current_version).is_none() {
+            return Err(format!(
+                "Current DB version {current_version} is not known"
+            ));
+        }
+
+        let target_versions = all_versions.range((
+            Bound::Excluded(&current_version),
+            Bound::Included(&desired_version),
+        ));
+
+        for target_version in target_versions.into_iter() {
+            info!(
+                log,
+                "Attempting to grab lease for upgrade to version {}",
+                target_version
+            );
+
+            // Before we attempt to perform an update, grab a lease to allow
+            // this Nexus instance to perform the update.
+            //
+            // If another instance is performing an update, we will fail, and
+            // let them to the work on our behalf.
+            let lease = self
+                .grab_schema_update_lease(&current_version, &target_version)
+                .await
+                .map_err(|e| e.to_string())?;
+
+            info!(
+                log,
+                "Acquired lease, applying upgrade to version {}",
+                target_version
+            );
+
+            // Start the schema change to the new version.
+            let up = config
+                .schema_dir
+                .join(target_version.to_string())
+                .join("up.sql");
+            let sql = tokio::fs::read_to_string(&up).await.map_err(|e| {
+                format!("Cannot read {up}: {e}", up = up.display())
+            })?;
+            self.apply_schema_update(&lease, &sql)
+                .await
+                .map_err(|e| e.to_string())?;
+
+            // NOTE: We could execute the schema change in a background task,
+            // and let it propagate, while observing it with the following
+            // snippet of SQL:
+            //
+            // WITH
+            //   x AS (SHOW JOBS)
+            // SELECT * FROM x WHERE
+            //   job_type = 'SCHEMA CHANGE' AND
+            //   status != 'succeeded';
+            //
+            // This would enable concurrent operations to happen on the database
+            // while we're mid-update. However, there is subtlety here around
+            // the visibility of renamed / deleted fields, unique indices, etc,
+            // so in the short-term we simply block on this job performing the
```

Comment:
"short-term" seems optimistic here, but optimism is a value!

Reply:
On a geological scale...